p2p/dial.go

```diff
@@ -0,0 +1,130 @@
+package p2p
+
+import (
+	"crypto/rand"
+	"fmt"
+	"net"
+
+	"github.com/ethereum/go-ethereum/logger"
+	"github.com/ethereum/go-ethereum/logger/glog"
+	"github.com/ethereum/go-ethereum/p2p/discover"
+)
+
+// dialstate creates dials and discovery lookups.
+type dialstate struct {
+	lookupBuf     []*discover.Node // current discovery lookup results
+	lookupRunning bool
+
+	static      map[discover.NodeID]*discover.Node
+	dialing     map[discover.NodeID]connFlag
+	maxDynDials int
+}
+
+func newDialState(static []*discover.Node, maxdyn int) *dialstate {
+	s := &dialstate{
+		maxDynDials: maxdyn,
+		static:      make(map[discover.NodeID]*discover.Node),
+		dialing:     make(map[discover.NodeID]connFlag),
+	}
+	for _, n := range static {
+		s.static[n.ID] = n
+	}
+	return s
+}
+
+func (s *dialstate) addStatic(n *discover.Node) {
+	s.static[n.ID] = n
+}
+
+func (s *dialstate) newTasks(peers map[discover.NodeID]*Peer) []task {
+	// Compute number of dynamic dials necessary at this point.
+	needDynDials := s.maxDynDials
+	for _, p := range peers {
+		if p.rw.is(dynDialedConn) {
+			needDynDials--
+		}
+	}
+	for _, flag := range s.dialing {
+		if flag&dynDialedConn != 0 {
+			needDynDials--
+		}
+	}
+
+	var newtasks []task
+	addDial := func(flag connFlag, n *discover.Node) {
+		_, inprogress := s.dialing[n.ID]
+		if peers[n.ID] == nil && !inprogress {
+			s.dialing[n.ID] = flag
+			newtasks = append(newtasks, &dialTask{flags: flag, dest: n})
+		}
+	}
+	// Create dials for static nodes if they are not connected.
+	for _, n := range s.static {
+		addDial(staticDialedConn, n)
+	}
+	// Create dynamic dials if lookup results are available
+	// and launch a lookup if they're not.
+	if needDynDials > 0 {
+		if len(s.lookupBuf) == 0 {
+			if !s.lookupRunning {
+				newtasks = append(newtasks, &discoverTask{})
+				s.lookupRunning = true
+			}
+		} else {
+			for i := 0; i < len(s.lookupBuf) && i <= needDynDials; i++ {
+				addDial(dynDialedConn, s.lookupBuf[i])
+			}
+			// TODO: consider keeping unused results.
+			s.lookupBuf = s.lookupBuf[:0]
+		}
```

Comment:
Maybe we could do lookups and exhaustion concurrently? Currently the dialer does a lookup, fills a buffer, then dials until it's depleted. A potential issue I see with this is that after it's depleted, the dialer needs to wait for a potentially long/slow/expensive lookup before it has something new to dial.

What we could do instead, is to run the lookup and dials concurrently:
```
if len(s.lookupBuf) <= needDynDials { // Note, equality is a nice pre-caching mechanism
  // Lookup
}
// No else needed, the check if the for already covers if the buffer is empty
for i := [...]
```

Along the same lines, I wouldn't discard the unused lookups. Keep them for next time as a lot of the dials will probably fail. Similarly, maybe we don't need the outer `if`. If the needed dynamic dials is 0, the for will skip anyway, but at least we'll prefetch some potential peers (`len(s.lookupBuf) == needDynDials == 0).